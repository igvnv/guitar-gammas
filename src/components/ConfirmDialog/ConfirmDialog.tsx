import React, { useState, useEffect } from 'react';
import classNames from 'classnames';

type ConfirmDialogProps = {
  /** Confirm dialog message */
  message: string;
  /** Callback, that will be fired when YES button is clicked */
  onConfirm: () => void;
  /** Callback, that will be fired when NO button is clicked */
  onCancel: () => void;
  /** "Yes" button text */
  yesButton?: string;
  /** "No" button text */
  noButton?: string;
  /** Time for opening/closing the dialog */
  animationDuration?: number;
};

/**
 * Confirm dialog.
 */
const ConfirmDialog: React.FC<ConfirmDialogProps> = ({
  message,
  onConfirm,
  onCancel,
  yesButton = 'Да',
  noButton = 'Нет',
  animationDuration = 400
}) => {
  const [result, setResult] = useState<boolean>();
  const [closing, setClosing] = useState(false);

  useEffect(() => {
    if (result === undefined) return;
    setClosing(true);

    const timeout = setTimeout(() => {
      if (result) {
        onConfirm();
      } else {
        onCancel();
      }
    }, animationDuration);

    return () => {
      clearTimeout(timeout);
    };
  }, [result, onCancel, onConfirm, animationDuration]);

  return (
    <>
      <div
        className={classNames('popup-background', {
          'popup-background_closing': closing
        })}
        style={{ animationDuration: `${animationDuration}ms` }}
      />
      <div
        className={classNames('alert', { alert_closing: closing })}
        style={{ animationDuration: `${animationDuration}ms` }}
      >
        <p className="alert__message">{message}</p>

        <div className="alert__buttons-list">
          <button
            type="button"
            className="alert__button alert__button_main"
            onClick={setResult.bind(null, false)}
            data-testid="no-button"
          >
            {noButton}
          </button>
          <button
            type="button"
            className="alert__button"
            onClick={setResult.bind(null, true)}
            data-testid="yes-button"
          >
            {yesButton}
          </button>
        </div>
      </div>
    </>
  );
};

export default ConfirmDialog;
