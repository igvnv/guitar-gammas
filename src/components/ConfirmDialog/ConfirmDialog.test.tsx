import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ConfirmDialog from './ConfirmDialog';

jest.useFakeTimers();

describe('ConfirmDialog', () => {
  it('displays message', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { container, getByText } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    getByText('Confirm message');
    expect(container).toMatchSnapshot();
  });

  it('has default text for YES button', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    expect(getByTestId('yes-button').innerHTML).toEqual('Да');
  });

  it('has default text for NO button', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    expect(getByTestId('no-button').innerHTML).toEqual('Нет');
  });

  it('can customize YES button', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        yesButton="Yes!"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    expect(getByTestId('yes-button').innerHTML).toEqual('Yes!');
  });

  it('has default text for NO button', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    expect(getByTestId('no-button').innerHTML).toEqual('Нет');
  });

  it('can customize NO button', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        noButton="No!"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    expect(getByTestId('no-button').innerHTML).toEqual('No!');
  });

  it('calls onConfirm when YES button is clicked', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    fireEvent.click(getByTestId('yes-button'));
    expect(onConfirm).toBeCalledTimes(0);

    jest.runAllTimers();

    expect(onConfirm).toBeCalledTimes(1);
    expect(onCancel).toBeCalledTimes(0);
  });

  it('calls onCancel when NO button is clicked', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
      />
    );

    fireEvent.click(getByTestId('no-button'));
    expect(onCancel).toBeCalledTimes(0);

    jest.runAllTimers();

    expect(onConfirm).toBeCalledTimes(0);
    expect(onCancel).toBeCalledTimes(1);
  });

  it('calls onConfirm after timeout', () => {
    const onCancel = jest.fn();
    const onConfirm = jest.fn();
    const { getByTestId } = render(
      <ConfirmDialog
        message="Confirm message"
        onCancel={onCancel}
        onConfirm={onConfirm}
        animationDuration={2000}
      />
    );

    fireEvent.click(getByTestId('no-button'));

    jest.runTimersToTime(1900);
    expect(onCancel).toBeCalledTimes(0);

    jest.runTimersToTime(100);
    expect(onCancel).toBeCalledTimes(1);
  });
});
