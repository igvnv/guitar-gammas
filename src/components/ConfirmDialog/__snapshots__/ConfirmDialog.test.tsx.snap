// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmDialog displays message 1`] = `
<div>
  <div
    class="popup-background"
    style="animation-duration: 400ms;"
  />
  <div
    class="alert"
    style="animation-duration: 400ms;"
  >
    <p
      class="alert__message"
    >
      Confirm message
    </p>
    <div
      class="alert__buttons-list"
    >
      <button
        class="alert__button alert__button_main"
        data-testid="no-button"
        type="button"
      >
        Нет
      </button>
      <button
        class="alert__button"
        data-testid="yes-button"
        type="button"
      >
        Да
      </button>
    </div>
  </div>
</div>
`;
